lib/engine/game/g_1860.rb

```diff
@@ -356,6 +359,175 @@ def legal_tile_rotation?(_entity, _hex, tile)
+      end
+
+      def check_hex_reentry(route)
+        visited_hexes = {}
+        last_hex = nil
+        route.ordered_paths.each do |path|
+          hex = path.hex
+          game_error('Route cannot re-enter a hex') if hex != last_hex && visited_hexes[hex]
+
+          visited_hexes[hex] = true
+          last_hex = hex
+        end
+      end
+
+      def check_other(route)
+        check_hex_reentry(route)
+      end
+
+      def max_halts(route)
+        # FIXME: need to ignore halts after formation of Southern Railway
+        visits = route.visited_stops
+        return 0 if visits.empty?
+
+        cities = visits.select { |node| node.city? || node.offboard? }
+        halts = visits.select(&:halt?)
+        th_allowance = route.train.distance.last + route.train.distance.first - cities.size
+        [halts.size, th_allowance].min
+      end
+
+      def compute_stops(route)
+        # FIXME: need to ignore halts after formation of Southern Railway
+        # will need to be modifed for original rules option
+        visits = route.visited_stops
+        return [] if visits.empty?
+
+        # no choice about citys/offboards => they must be stops
+        stops = visits.select { |node| node.city? || node.offboard? }
+
+        # in 1860, unused city/offboard allowance can be used for towns/halts
+        c_allowance = route.train.distance.first
+        th_allowance = route.train.distance.last + c_allowance - stops.size
+
+        # add in halts requested
+        halts = visits.select(&:halt?)
+        num_halts = [halts.size, (route.num_halts || 0)].min
+        if num_halts.positive?
+          stops.concat(halts.take(num_halts))
+          th_allowance -= num_halts
+        end
+
+        # pick highest revenue towns
+        towns = visits.select { |node| node.town? && !node.halt? }
+        num_towns = [th_allowance, towns.size].min
+        if num_towns.positive?
+          stops.concat(towns.sort_by(&:revenue).take(num_towns))
+          th_allowance -= num_towns
+        end
+
+        # if this is first time for this route, add as many halts as possible
+        if !route.num_halts && halts.any? && th_allowance.positive?
+          num_halts = [halts.size, th_allowance].min
+          stops.concat(halts.take(num_halts))
+        end
+
+        route.num_halts = num_halts if halts.any?
+
+        game_error('Logic bug! too many stops selected for revenue') if stops.size > route.train.distance.sum
+
+        stops
+      end
+
+      def route_distance(route)
+        n_cities = route.stops.select { |n| n.city? || n.offboard? }.size
+        n_towns = route.stops.select { |n| n.town? && !n.halt? }.size
+        "#{n_cities}+#{n_towns}"
+      end
+
+      def revenue_for(route, stops)
+        stops.sum { |stop| stop.route_base_revenue(route.phase, route.train) }
+      end
+
+      def subsidy_for(_route, stops)
+        stops.select(&:halt?).size * HALT_SUBSIDY
```

Comment:
stops.count(&:halt?)

Reply:
but how does this work with num_halts in route?

Reply:
route.num_halts is used by @game.compute_halts to determine how many stops are at halts.

Reply:
doesn't that mean that this method should check num_halts and not select halts? don't you need to know which ones the user chose?

Reply:
All halts are the same, I just need to know how many halts are being stopped at. I could use either Route::halts or count the stops that are halts. They should always be the same.